### PR TITLE
vmm: assign each pci segment 32-bit mmio allocator

### PR DESCRIPTION
**Pull request by thomasbarrett**
This PR is a fix for https://github.com/cloud-hypervisor/cloud-hypervisor/issues/5952. I have followed the pattern established for 64-bit memory regions to creating a separate 32-bit memory area allocator for each PCI segment (as opposed to assigning the entire 32-bit memory area to PCI segment 0). This allows passthrough of VFIO devices w/ 32-bit memory BARs to non-zero PCI segments.

### CODE REVIEW

**Comment by thomasbarrett** [comment]
Update: it works!  Shown below is a NVIDIA A100 GPU w/ a 32-bit Memory BAR running on a non-zero PCI segment. 
```
lspci -s 0001:00:05.0 -vvv
0001:00:05.0 3D controller: NVIDIA Corporation GA100 [A100 PCIe 80GB] (rev a1)
	Subsystem: NVIDIA Corporation GA100 [A100 PCIe 80GB]
	Physical Slot: 5-2
	Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Latency: 0
	Interrupt: pin A routed to IRQ 125
	NUMA node: 0
	Region 0: Memory at d9000000 (32-bit, non-prefetchable) [size=16M]
	Region 1: Memory at 2ab23fe00000 (64-bit, prefetchable) [size=128G]
	Region 3: Memory at 2ab23de00000 (64-bit, prefetchable) [size=32M]
```

File: vmm/src/device_manager.rs
Comment (jinankjain) [inline vmm/src/device_manager.rs:627]: nit: You can use the defined constant PVPANIC_DEVICE_MMIO_ALIGNMENT

File: vmm/src/device_manager.rs
Comment (jinankjain) [inline vmm/src/device_manager.rs:657]: Same here use of defined constant PVPANIC_DEVICE_MMIO_ALIGNMENT

File: vmm/src/device_manager.rs
Comment (rbradford) [reply vmm/src/device_manager.rs:627]: No, I don't think that is an appropriate use of that constant.

File: vmm/src/device_manager.rs
Comment (rbradford) [reply vmm/src/device_manager.rs:657]: Ditto.

File: vmm/src/device_manager.rs
Comment (rbradford) [reply vmm/src/device_manager.rs:627]: I'm not sure amount the comment (perhaps it should say length rather than alignment) as this sets a minimum length. I think the PCI bar must be at least 4KiB aligned.

**Review by rbradford — COMMENTED** [review]
Good, work. If you use the maximum number of PCI segments is the size of the 32-bit space still usable - that was one of the reasons I only implemented 64-bit bar support.

File: vmm/src/device_manager.rs
Comment (thomasbarrett) [reply vmm/src/device_manager.rs:627]: Good point. After a closer reading, it looks like BARs must be naturally aligned, and the minimum BAR size is 0x10. With a 2-byte BAR `pv_panic` device is a bit out-of-spec (although it appears to work just fine anyway). For VFIO devices, there is an additional page-size alignment requirement that doesn't apply to emulated devices. I'm going to remove these two changes and focus on the memory allocator (but keep the change to the pv_panic device since that is sufficient to allow integration tests to pass).

**Comment by thomasbarrett** [comment]
> Good, work. If you use the maximum number of PCI segments is the size of the 32-bit space still usable - that was one of the reasons I only implemented 64-bit bar support.

Good point. With the current maximum of 96 PCI segments and the current 32-bit memory hole size of 640MiB, each segment would get ~6MiB of 32-bit memory space. This really isn't all that much, but it is usable for a handful of legacy devices. We could also:
1. Give segment 0 a greater share of the 640 MiB space than the rest of the segments. We could configure something like 40 MiB minimum for segment 0 and `600 / pci_segments` for each segment.
2. Compute the amount of 32-bit space needed per segment based on arguments and use it to inform the distribution of 32-bit memory between segments. This wouldn't account for hot-plug devices.
2. The most generic solution would be to allow users to set the 32-bit and 64-bit aperture sizes as an argument when creating a virtual machine.

I think that any of these solutions should be considered once users starts running into the limitations of the simple allocation scheme. I imagine that most uses cases of PCI segments would only use ~8-16 segments max (40-80 MiB 32-bit memory per segment... a lot more reasonable).

**Review by rbradford — COMMENTED** [review]
@thomasbarrett Looking good - can you split pvpanic change into an earlier commit. It's a distinct bug fix that we might want to cherry pick.

**Comment by thomasbarrett** [comment]
> @thomasbarrett Looking good - can you split pvpanic change into an earlier commit. It's a distinct bug fix that we might want to cherry pick.

I realize now that you probably meant that I should split this PR into two commits. I just opened a second PR with the pv_panic commit.

**Comment by rbradford** [comment]
@thomasbarrett Please can you rebase.

**Review by rbradford — APPROVED** [review]
Good work!

**Comment by thomasbarrett** [comment]
Should I rebase this @rbradford? Sorry, I'm not entirely sure what the protocol is for merging approved PRs if I don't have write access.

**Comment by rbradford** [comment]
> Should I rebase this @rbradford? Sorry, I'm not entirely sure what the protocol is for merging approved PRs if I don't have write access.

I'm giving others e.g. @likebreath an opportunity to review if they wish.

**Comment by thomasbarrett** [comment]
> > Should I rebase this @rbradford? Sorry, I'm not entirely sure what the protocol is for merging approved PRs if I don't have write access.
> 
> I'm giving others e.g. @likebreath an opportunity to review if they wish.

Makes sense. Thats what I thought. Just making sure you weren't waiting on action from me :).

File: vm-allocator/src/system.rs
Comment (likebreath) [inline vm-allocator/src/system.rs:72]: Nit: you forgot to remove the comments about the deleted inputs.

File: vmm/src/device_manager.rs
Comment (likebreath) [inline vmm/src/device_manager.rs:638]: We are now doing the same thing for 32-bit and 64-bit MMIO regions, so we can further simplify the code like:

```
                let allocators = if region_type == PciBarRegionType::Memory32BitRegion {                                                                                                                      
                    &self.pci_mmio32_allocators                                                                                                                                                               
                } else {                                                                                                                                                                                      
                    &self.pci_mmio64_allocators                                                                                                                                                               
                };   

                // Find the specific allocator that this BAR was allocated from and use it for new one                                                                                                        
                for allocator in allocators { 
```

File: vmm/src/device_manager.rs
Comment (likebreath) [inline vmm/src/device_manager.rs:1052]: I believe it is cleaner to use a "closure" like below: 

```
let create_allocators = |start, end, num_pci_segments: u16, aligment: u64| {};
```

**Review by likebreath — COMMENTED** [review]
Overall look good. Couple comments on improving readability. Thank you.

File: vmm/src/device_manager.rs
Comment (thomasbarrett) [reply vmm/src/device_manager.rs]: Ack. Love it. A lot less repetitive. Updated.

File: vmm/src/device_manager.rs
Comment (thomasbarrett) [reply vmm/src/device_manager.rs]: This is a lot less repetitive. Updated.

File: vm-allocator/src/system.rs
Comment (thomasbarrett) [reply vm-allocator/src/system.rs:72]: Updated.

**Review by likebreath — APPROVED** [review]

**Pull request merged**
